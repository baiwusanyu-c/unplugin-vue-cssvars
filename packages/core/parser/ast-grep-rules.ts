export const astGrepRules = {
  IDENT_NAME: {
    has: {
      kind: 'identifier',
      field: 'name',
    },
  },
  FN_CALL: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'call_expression',
      },
    },
  },
  ARROW_FN: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'arrow_function',
      },
    },
  },
  NOR_FN_VAR: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'function',
      },
    },
  },
  OBJ_VAR: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'object',
        field: 'value',
      },
    },
  },
  ARR_VAR: {
    has: {
      kind: 'variable_declarator',
      has: {
        kind: 'array',
        field: 'value',
      },
    },
  },
  NOR_FN: {
    has: {
      kind: 'function_declaration',
      has: {
        kind: 'identifier',
        field: 'name',
      },
    },
  },
  PROPS_DEFAULT_CALL: {
    has: {
      kind: 'call_expression',
      has: {
        kind: 'identifier',
        field: 'function',
      },
    },
  },
  PROPS_DEFAULT_ARG: {
    kind: 'property_identifier',
    inside: {
      kind: 'pair',
      inside: {
        kind: 'object',
        inside: {
          kind: 'arguments',
        },
      },
    },
  },
  PROPS_DEFAULT_VAL: {
    kind: 'object',
    inside: {
      kind: 'arguments',
    },
  },
  CONST_VAR: {
    any: [
      {
        pattern: 'const $VAR',
      },
    ],
  },
  LET_VAR: {
    any: [
      {
        pattern: 'let $VAR',
      },
    ],
  },
  CONST_REF_VAR: {
    any: [
      {
        pattern: 'const $VAR = ref($VAL)',
      },
    ],
  },
  CONST_REACTIVE_VAR: {
    any: [
      {
        pattern: 'const $VAR = reactive($VAL)',
      },
    ],
  },
  CONST_PROPS_VAR: {
    any: [
      {
        pattern: 'const $VAR = defineProps($VAL)',
      },
    ],
  },
  CONST_NOR_FN: {
    has:
      {
        kind: 'function_declaration',
      },
  },
  OBJ_PATTERN: {
    has:
      {
        kind: 'object_pattern',
      },
  },
  NEW_EXP: {
    has:
      {
        kind: 'new_expression',
        field: 'value'
      },
  },
  OBJ_PATTERN_VAL: {
    any:
      [
        {
          kind: 'identifier',
        },
        {
          kind: 'shorthand_property_identifier_pattern',
        },
      ],
  },
}

export const getRules = (name: string) => {
  return {
    rule: {
      matches: name,
    },
    utils: astGrepRules,
  }
}
