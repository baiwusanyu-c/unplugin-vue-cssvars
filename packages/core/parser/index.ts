export * from './parser-variable'
export * from './parser-import'
export * from './parser-vbindm'
export * from './parser-compiled-sfc'
