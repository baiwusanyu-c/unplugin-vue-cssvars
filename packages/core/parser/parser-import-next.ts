export enum ParserState {
  Initial,
  At,
  AtImport,
  AtUse,
  AtRequire,
  StringLiteral,
}

export interface ImportStatement {
  type: 'import' | 'use' | 'require'
  path: string
  start?: number
  end?: number
  suffix?: string
}
const delTransformSymbol = (content: string) => content.replace(/[\r\t\f\v\\]/g, '')
export function parseImportsNext(content: string): {
  imports: ImportStatement[]
  getCurState: () => ParserState
  getCurImport: () => undefined | ImportStatement
} {
  const imports: ImportStatement[] = []
  let currentImport: ImportStatement | undefined
  let state = ParserState.Initial
  let i = 0
  let AtPath = ''
  const source = delTransformSymbol(content)
  while (i < source.length) {
    const char = source[i]
    switch (state) {
      case ParserState.Initial:
        if (char === '@')
          state = ParserState.At
        break
      case ParserState.At:
        AtPath = AtPath + char
        if (AtPath === 'import') {
          AtPath = ''
          state = ParserState.AtImport
          currentImport = { type: 'import', path: '' }
          i++ // skip over "i" to next character
        } else if (AtPath === 'use') {
          AtPath = ''
          state = ParserState.AtUse
          currentImport = { type: 'use', path: '' }
          i++ // skip over "u" to next character
        } else if (AtPath === 'require') {
          AtPath = ''
          state = ParserState.AtRequire
          currentImport = { type: 'require', path: '' }
          i++ // skip over "u" to next character
        }
        // TODO ...
        if (AtPath.length >= 7 && AtPath !== 'require')
          state = ParserState.Initial

        break
      case ParserState.AtImport:
      case ParserState.AtUse:
      case ParserState.AtRequire:
        debugger
        // 当字符不是空格，且前一个是空格，进入取值
        if (char !== ' ' && source[i - 1] === ' ') {
          state = ParserState.StringLiteral
          currentImport!.start = i
          currentImport!.path += char
        } else if (char === '\n' || i === source.length - 1) {
          if (currentImport && currentImport.start !== undefined) {
            currentImport.end = i
            imports.push(currentImport)
            currentImport = undefined
          }
          state = ParserState.Initial
        }
        break
      case ParserState.StringLiteral:
        // 遇到引号，其起始位置也是，则是有引号状态下的取值结束
        if ((char === "'" || char === '"')
          && (currentImport!.start || currentImport!.start === 0)
          && (source[currentImport!.start] === char)) {
          if (currentImport!.type === 'import')
            state = ParserState.AtImport

          if (currentImport!.type === 'use')
            state = ParserState.AtUse

          if (currentImport!.type === 'require')
            state = ParserState.AtRequire

          currentImport!.path += char
        } else {
          // 取值
          currentImport!.path += char
        }

        break
    }
    i++
  }

  function getCurState() {
    return state
  }
  function getCurImport() {
    return currentImport
  }
  return {
    imports,
    getCurState,
    getCurImport,
  }
}
