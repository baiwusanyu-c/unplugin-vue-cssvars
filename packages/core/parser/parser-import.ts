const innerAtRule = 'media,extend,at-root,debug,warn,forward,mixin,include,function,error,keyframes,font-face,page,supports,namespace,return,if,else,for,while,each,content'
export enum ParserState {
  Initial,
  InlineComment,
  Comment,
  AtStart,
  AtEnd,
  AtImport,
  AtUse,
  AtRequire,
  QuotesStart,
  QuotesEnd,
  StringLiteral,
}

export interface ImportStatement {
  type: 'import' | 'use' | 'require'
  path: string
  start?: number
  end?: number
  suffix?: string
}

export function parseImports(content: string, helper?: Array<Function>): {
  imports: ImportStatement[]
  getCurState: () => ParserState
  getCurImport: () => undefined | ImportStatement
} {
  const imports: ImportStatement[] = []
  let currentImport: ImportStatement | undefined
  let state = ParserState.Initial
  let i = 0
  let AtPath = ''
  const source = content
  while (i < source.length) {
    const char = source[i]
    if (/[\r\t\f\v\\]/g.test(char)) {
      i++
      continue
    }

    switch (state) {
      case ParserState.Initial:
        if (char === '@')
          state = ParserState.AtStart
        if (char === '/' && source[i + 1] === '/')
          state = ParserState.InlineComment
        if (char === '/' && source[i + 1] === '*')
          state = ParserState.Comment
        break
      case ParserState.InlineComment:
        if (char === '\n')
          state = ParserState.Initial
        break
      case ParserState.Comment:
        if (char === '*' && source[i + 1] === '/')
          state = ParserState.Initial
        break
      case ParserState.AtStart:
        if (/[A-Za-z]$/.test(char))
          AtPath = AtPath + char
        else
          state = ParserState.AtEnd

        if (i === source.length - 1) {
          if (char === '"' || char === "'")
            throw new Error('syntax error: unmatched quotes')
          else
            walkContentEnd(i)
        }
        if (!(/[A-Za-z]$/.test(char))
          && char !== '\n'
          && char !== ' '
          && char !== '-')
          throw new Error('syntax error')

        break
      case ParserState.AtEnd:
        if (char !== '\n' && char !== ' ' && char !== '-') {
          if (char === '/')
            throw new Error('syntax error')

          if (AtPath === 'import') {
            AtPath = ''
            state = ParserState.AtImport
            currentImport = { type: 'import', path: '' }
            i--
          } else if (AtPath === 'use') {
            AtPath = ''
            state = ParserState.AtUse
            currentImport = { type: 'use', path: '' }
            i--
          } else if (AtPath === 'require') {
            AtPath = ''
            state = ParserState.AtRequire
            currentImport = { type: 'require', path: '' }
            i--
          } else {
            if (!innerAtRule.includes(AtPath))
              throw new Error('syntax error: unknown At Rule')

            AtPath = ''
            state = ParserState.Initial
          }
        }

        break
      case ParserState.AtImport:
      case ParserState.AtUse:
      case ParserState.AtRequire:
        // '@require test.css;@require test2.css'
        if (char === '@' && !(/[A-Za-z]$/.test(source[i - 1]))) {
          i--
          state = ParserState.Initial
          break
        }

        if (char === '/' && (source[i + 1] === '/' || source[i + 1] === '*')) {
          i--
          state = ParserState.Initial
          break
        }

        if (char === "'" || char === '"') {
          currentImport!.start = i
          state = ParserState.QuotesStart
          break
        }
        if (char !== '\n' && char !== ' ') {
          currentImport!.start = i
          currentImport!.path += char
          state = ParserState.StringLiteral
          break
        }
        break
      case ParserState.QuotesStart:
        if (char === "'" || char === '"') {
          currentImport!.end = i
          state = ParserState.QuotesEnd
          if (i === source.length - 1)
            walkContentEnd(i)
          break
        }
        if (i === source.length - 1)
          throw new Error('syntax error: unmatched quotes')

        currentImport!.path += char
        break
      case ParserState.QuotesEnd:
        if (i === source.length - 1 || char === ';' || char === '\n') {
          walkContentEnd(i)
        } else {
          i--
          state = ParserState.StringLiteral
        }
        break
      case ParserState.StringLiteral:
        if (char === ';' || char === '\n') {
          walkContentEnd(i)
          break
        }

        // '@require test.css@require test2.css'
        // '@import ./test1, ./test2'
        if (
          char !== '@'
          && (char === ' '
          || char === ','
          || char === '"'
          || char === "'")) {
          const curType = currentImport?.type
          walkContentEnd(i)
          if (curType === 'import') {
            state = ParserState.AtImport
            currentImport = { type: 'import', path: '' }
          } else if (curType === 'use') {
            state = ParserState.AtUse
            currentImport = { type: 'use', path: '' }
          } else if (curType === 'require') {
            state = ParserState.AtRequire
            currentImport = { type: 'require', path: '' }
          }

          if (char === "'" || char === '"') {
            currentImport!.start = i
            state = ParserState.QuotesStart
            if (i === source.length - 1 && (char === '"' || char === "'"))
              throw new Error('syntax error: unmatched quotes')

            break
          }

          break
        }

        currentImport!.path += char
        if (i === source.length - 1)
          walkContentEnd(i)

        break
    }
    i++
  }

  function walkContentEnd(index: number) {
    pushCurrentImport(index)
    state = ParserState.Initial
  }

  function pushCurrentImport(index: number) {
    if (currentImport && currentImport.start !== undefined) {
      currentImport.end = index
      if (helper) {
        helper.forEach((fn) => {
          currentImport = fn(currentImport)
        })
      }
      imports.push(currentImport)
      currentImport = undefined
    }
  }

  function getCurState() {
    return state
  }
  function getCurImport() {
    return currentImport
  }
  return {
    imports,
    getCurState,
    getCurImport,
  }
}
