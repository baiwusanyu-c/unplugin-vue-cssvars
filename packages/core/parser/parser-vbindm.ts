
export enum ParserVBindMState {
  Initial,
  InlineComment,
  Comment,
  VBindM,
  VBindMValueStart,
  VBindMValue,
  VBindMValueEnd,
  StringLiteral,
}

export function parseVBindM(source: string): Array<string> {
  const result = new Set<string>()
  let state = ParserVBindMState.Initial
  let i = 0
  let curBuffer = ''
  let mark = 0

  while (i < source.length) {
    const char = source[i]
    switch (state) {
      case ParserVBindMState.Initial:
        if (char === '/' && source[i + 1] === '/')
          state = ParserVBindMState.InlineComment
        if (char === '/' && source[i + 1] === '*')
          state = ParserVBindMState.Comment
        if (char === 'v')
          state = ParserVBindMState.VBindM
        if (char === ')')
          mark--
        if (char === '(')
          mark++
        break
      case ParserVBindMState.InlineComment:
        if (char === '\n')
          state = ParserVBindMState.Initial
        break
      case ParserVBindMState.Comment:
        if (char === '*' && source[i + 1] === '/')
          state = ParserVBindMState.Initial
        if (i === source.length - 1)
          throw new Error('syntax error: unmatched */')
        break
      case ParserVBindMState.VBindM:
        if (char === '-'
            && source[i + 1] === 'b'
            && source[i + 2] === 'i'
            && source[i + 3] === 'n'
            && source[i + 4] === 'd'
            && source[i + 5] === '-'
            && source[i + 6] === 'm') {
          state = ParserVBindMState.VBindMValueStart
          i = i + 6
        }
        break
      case ParserVBindMState.VBindMValueStart:
        if (char !== '(') {
          i--
          state = ParserVBindMState.VBindMValue
          break
        }

        mark++
        break
      case ParserVBindMState.VBindMValue:
        if (char !== ')') {
          curBuffer = curBuffer + char
        } else {
          i--
          state = ParserVBindMState.VBindMValueEnd
        }
        break
      case ParserVBindMState.VBindMValueEnd:
        if (char === ')') {
          mark--
          if (curBuffer) {
            result.add(curBuffer)
            curBuffer = ''
          }
          state = ParserVBindMState.Initial
        }
        break
    }
    i++
  }

  if (mark > 0)
    throw new Error('syntax error: unmatched )')
  if (mark < 0)
    throw new Error('syntax error: unmatched (')

  return [...result]
}
