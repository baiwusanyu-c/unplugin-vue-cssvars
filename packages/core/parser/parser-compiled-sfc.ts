import { parse as babelParse } from '@babel/parser'
import { walk } from 'estree-walker-ts'
import type {
  BlockStatement,
  CallExpression, ExpressionStatement,
  Identifier,
  ImportSpecifier,
  ObjectExpression,
} from '@babel/types'
import type { Node } from 'estree-walker-ts'

export interface IParseSFCRes {
  setupBodyNode: BlockStatement
  hasCSSVars: boolean
  useCSSVarsNode: ObjectExpression
}

let isSetupEnter = false
let setupBodyNode = {} as BlockStatement
// TODO: unit test
export function parseSetupBody(node: Node & { scopeIds?: Set<string> }) {
  if ((node as Identifier).type === 'Identifier' && (node as Identifier).name === 'setup') {
    isSetupEnter = true
    return
  }
  if (isSetupEnter && (node as BlockStatement).type === 'BlockStatement') {
    isSetupEnter = false
    setupBodyNode = node as BlockStatement
  }
}

let hasCSSVars = false
// TODO: unit test
export function parseHasCSSVars(
  node: Node & { scopeIds?: Set<string> },
  parent: ImportSpecifier) {
  if ((node as Identifier).type === 'Identifier'
    && (node as Identifier).name === 'useCssVars'
    && parent
    && ((parent.type as ImportSpecifier) === 'ImportSpecifier'))
    hasCSSVars = true
}

let useCSSVarsNode = {} as ExpressionStatement
let isUseCSSVarsEnter = false
export function parseUseCSSVars(
  node: Node & { scopeIds?: Set<string> },
  parent: CallExpression) {
  if ((node as Identifier).type === 'Identifier'
    && (node as Identifier).name === '_useCssVars'
    && parent
    && ((parent.type as CallExpression) === 'CallExpression'))
    isUseCSSVarsEnter = true

  if (isUseCSSVarsEnter && (node as ObjectExpression).type === 'ObjectExpression') {
    isUseCSSVarsEnter = false
    useCSSVarsNode = node as ExpressionStatement
  }
}

// TODO: unit test
export function parserCompiledSfc(code: string) {
  isSetupEnter = false
  setupBodyNode = {} as BlockStatement
  isUseCSSVarsEnter = false
  useCSSVarsNode = {} as ExpressionStatement
  hasCSSVars = false

  const ast = babelParse(code, {
    sourceType: 'module',
    plugins: ['typescript'],
  });

  (walk as any)(ast, {
    enter(
      node: Node & { scopeIds?: Set<string> },
      parent: Node | undefined | ObjectExpression | ImportSpecifier | CallExpression) {
      parseSetupBody(node)
      parseHasCSSVars(node, parent as ImportSpecifier)
      parseUseCSSVars(node, parent as CallExpression)
    },
  })
  return {
    setupBodyNode,
    hasCSSVars,
    useCSSVarsNode,
  } as IParseSFCRes
}
