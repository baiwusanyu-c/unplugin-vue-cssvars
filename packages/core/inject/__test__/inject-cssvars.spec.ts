import { describe, expect, test } from 'vitest'
import MagicString from 'magic-string'
import {
  createCSSVarsObjCode,
  createUCVCOptionUnHas,
  createUCVCSetupUnHas,
  createUseCssVarsCode, injectCSSVars,
  injectCSSVarsOnServer,
  injectUseCssVarsOption,
  injectUseCssVarsSetup,
} from '../inject-cssvars'
import { parserCompiledSfc } from '../../parser'

describe('inject-cssvars', () => {
  describe('createUCVCOptionUnHas', () => {
    test('returns a string containing the injected CSS variables', () => {
      const resCode = 'color: var(--text-color); font-size: var(--font-size);'
      const result = createUCVCOptionUnHas(resCode)
      expect(result).toContain('__injectCSSVars__')
      expect(result).toContain('useCssVars')
      expect(result).toContain(resCode)
    })
  })

  describe('createUCVCSetupUnHas', () => {
    test('returns a string containing the injected CSS variables', () => {
      const cssvarsObjectCode = '{ "--text-color": "red", "--font-size": "16px" }'
      const result = createUCVCSetupUnHas(cssvarsObjectCode)
      expect(result).toContain('useCssVars')
      expect(result).toContain(cssvarsObjectCode)
    })
  })

  describe('createUseCssVarsCode', () => {
    test('returns a string containing the injected CSS variables for script setup', () => {
      const cssvarsObjectCode = '{ "--text-color": "red", "--font-size": "16px" }'
      const isScriptSetup = true
      const expected = createUCVCSetupUnHas(cssvarsObjectCode)
      const result = createUseCssVarsCode(cssvarsObjectCode, isScriptSetup)
      expect(result).toBe(expected)
    })

    test('returns a string containing the injected CSS variables for composition api and option api', () => {
      const cssvarsObjectCode = '{ "--text-color": "red", "--font-size": "16px" }'
      const isScriptSetup = false
      const expected = createUCVCOptionUnHas(cssvarsObjectCode)
      const result = createUseCssVarsCode(cssvarsObjectCode, isScriptSetup)
      expect(result).toBe(expected)
    })
  })

  describe('createCSSVarsObjCode', () => {
    test('should return empty string if vbindVariableList is empty', () => {
      const vbindVariableList: any[] = []
      const result = createCSSVarsObjCode(vbindVariableList, false)
      expect(result).toBe('')
    })

    test('isScriptSetup=false, has=false, hash=false, isRef=false', () => {
      const vbindVariableList = [{ value: 'color', has: false, hash: false, isRef: false }]
      const result = createCSSVarsObjCode(vbindVariableList as any, false)
      expect(result).toContain('_ctx.color')
    })

    test('isScriptSetup=false, has=false, hash=false, isRef=true', () => {
      const vbindVariableList = [{ value: 'color', has: false, hash: false, isRef: true }]
      const result = createCSSVarsObjCode(vbindVariableList as any, false)
      expect(result).toContain('_ctx.color')
    })

    test('isScriptSetup=false, has=false, hash=true, isRef=false', () => {
      const vbindVariableList = [{ value: 'color', has: false, hash: 'color', isRef: false }]
      const result = createCSSVarsObjCode(vbindVariableList as any, false)
      expect(result).toContain('_ctx.color')
      expect(result).toContain('"color":')
    })

    test('isScriptSetup=false, has=false, hash=true, isRef=true', () => {
      const vbindVariableList = [{ value: 'color', has: false, hash: 'color', isRef: true }]
      const result = createCSSVarsObjCode(vbindVariableList as any, false)
      expect(result).toContain('_ctx.color')
      expect(result).toContain('"color":')
    })

    test('isScriptSetup=false, has=true, hash=false, isRef=false', () => {
      const vbindVariableList = [{ value: 'color', has: true, hash: false, isRef: false }]
      const result = createCSSVarsObjCode(vbindVariableList as any, false)
      expect(result).toContain('_ctx.color')
    })

    test('isScriptSetup=false, has=true, hash=false, isRef=true', () => {
      const vbindVariableList = [{ value: 'color', has: true, hash: false, isRef: true }]
      const result = createCSSVarsObjCode(vbindVariableList as any, false)
      expect(result).toContain('_ctx.color')
    })

    test('isScriptSetup=false, has=true, hash=true, isRef=false', () => {
      const vbindVariableList = [{ value: 'color', has: true, hash: 'color', isRef: false }]
      const result = createCSSVarsObjCode(vbindVariableList as any, false)
      expect(result).toContain('_ctx.color')
      expect(result).toContain('"color":')
    })

    test('isScriptSetup=false, has=true, hash=true, isRef=true', () => {
      const vbindVariableList = [{ value: 'color', has: true, hash: 'color', isRef: true }]
      const result = createCSSVarsObjCode(vbindVariableList as any, false)
      expect(result).toContain('_ctx.color')
      expect(result).toContain('"color":')
    })

    test('isScriptSetup=true, has=false, hash=false, isRef=false', () => {
      const vbindVariableList = [{ value: 'color', has: false, hash: false, isRef: false }]
      const result = createCSSVarsObjCode(vbindVariableList as any, true)
      expect(result).toContain('_ctx.color')
    })

    test('isScriptSetup=true, has=false, hash=false, isRef=true', () => {
      const vbindVariableList = [{ value: 'color', has: false, hash: false, isRef: true }]
      const result = createCSSVarsObjCode(vbindVariableList as any, true)
      expect(result).toContain('color.value')
    })

    test('isScriptSetup=true, has=false, hash=true, isRef=false', () => {
      const vbindVariableList = [{ value: 'color', has: false, hash: 'color', isRef: false }]
      const result = createCSSVarsObjCode(vbindVariableList as any, true)
      expect(result).toContain('_ctx.color')
      expect(result).toContain('"color":')
    })

    test('isScriptSetup=true, has=false, hash=true, isRef=true', () => {
      const vbindVariableList = [{ value: 'color', has: false, hash: 'color', isRef: true }]
      const result = createCSSVarsObjCode(vbindVariableList as any, true)
      expect(result).toContain('color.value')
      expect(result).toContain('"color":')
    })

    test('isScriptSetup=true, has=true, hash=false, isRef=false', () => {
      const vbindVariableList = [{ value: 'color', has: true, hash: false, isRef: false }]
      const result = createCSSVarsObjCode(vbindVariableList as any, true)
      expect(result).toContain('color')
    })

    test('isScriptSetup=true, has=true, hash=false, isRef=true', () => {
      const vbindVariableList = [{ value: 'color', has: true, hash: false, isRef: true }]
      const result = createCSSVarsObjCode(vbindVariableList as any, true)
      expect(result).toContain('color.value')
    })

    test('isScriptSetup=true, has=true, hash=true, isRef=false', () => {
      const vbindVariableList = [{ value: 'color', has: true, hash: 'color', isRef: false }]
      const result = createCSSVarsObjCode(vbindVariableList as any, true)
      expect(result).toContain('color,')
      expect(result).toContain('"color":')
    })

    test('isScriptSetup=true, has=true, hash=true, isRef=true', () => {
      const vbindVariableList = [{ value: 'color', has: true, hash: 'color', isRef: true }]
      const result = createCSSVarsObjCode(vbindVariableList as any, true)
      expect(result).toContain('color.value')
      expect(result).toContain('"color":')
    })

    test('avoid repeated injections', () => {
      const mockCode = '\n            "color": color.value,'
      const vbindVariableList = [{ value: 'color', has: true, hash: 'color', isRef: true }]
      const result = createCSSVarsObjCode(vbindVariableList as any, true, new MagicString(mockCode))
      expect(result).not.toBeTruthy()
    })
  })
  describe('injectUseCssVarsOption', () => {
    test('hasUseCssVars is false', () => {
      const mockContent = 'const _sfc_main = {};function _sfc_render(){}'
      const mgcStr = new MagicString(mockContent)
      const parserRes = parserCompiledSfc(mockContent)
      const useCssVars = 'const __injectCSSVars__ = () => {}'
      const result = injectUseCssVarsOption(
        mgcStr,
        useCssVars,
        false,
        parserRes,
      )
      expect(result.toString()).toMatchSnapshot()
    })

    test('hasUseCssVars is true', () => {
      const mockContent = 'import { useCssVars as _useCssVars } from \'vue\'\n'
        + 'const __injectCSSVars__ = () => {\n'
        + '_useCssVars(_ctx => ({\n'
        + '  "229090c3-sassColor": (_ctx.sassColor)\n'
        + '}))}\n'
        + 'const __setup__ = __default__.setup\n'
        + '__default__.setup = __setup__\n'
        + '  ? (props, ctx) => { __injectCSSVars__();return __setup__(props, ctx) }\n'
        + '  : __injectCSSVars__\n'
        + '\n'
        + 'const _sfc_main = __default__'
      const mgcStr = new MagicString(mockContent)
      const parserRes = parserCompiledSfc(mockContent)
      const useCssVars = '\n"color": (_ctx.color),\n'
      const result = injectUseCssVarsOption(
        mgcStr,
        useCssVars,
        true,
        parserRes,
      )
      expect(result.toString()).toMatchSnapshot()
    })
  })

  describe('injectUseCssVarsSetup', () => {
    test('hasUseCssVars is true', () => {
      const mockContent = 'function setup() {  '
        + '_useCssVars((_ctx) => ({\n'
        + '    "229090c3-sassColor": sassColor.value\n'
        + '})); }'
      const mgcStr = new MagicString(mockContent)
      const useCssVars = '"1439c43b": color.value,'
      const resMgcStr = injectUseCssVarsSetup(
        mgcStr,
        useCssVars,
        true,
        parserCompiledSfc(mockContent))
      expect(resMgcStr.toString()).toMatchSnapshot()
    })

    test('hasUseCssVars is false', () => {
      const mockContent = 'function setup() {}'
      const mgcStr = new MagicString(mockContent)
      const useCssVars = createUseCssVarsCode(
        '"1439c43b": color.value,',
        true)
      const resMgcStr = injectUseCssVarsSetup(
        mgcStr,
        useCssVars,
        false,
        parserCompiledSfc(mockContent))

      expect(resMgcStr.toString()).toMatchSnapshot()
    })
  })

  describe('injectCSSVarsOnServer', () => {
    test('isScriptSetup=false hasUseCssVars=false', () => {
      const mockContent = 'const _sfc_main = {};function _sfc_render(){}'
      const parseRes = parserCompiledSfc(mockContent)
      const mgcStr = new MagicString(mockContent)
      const res = injectCSSVarsOnServer(
        [{ value: 'color', has: false, isRef: true }] as any,
        false,
        parseRes,
        mgcStr)
      expect(res.mgcStr.toString()).toContain('_ctx.color')
      expect(res.mgcStr.toString()).toContain('_useCssVars')
      expect(res.mgcStr.toString()).toContain('_useCssVars')
    })

    test('isScriptSetup=true hasUseCssVars=true', () => {
      const mockContent = 'const _sfc_main = /* @__PURE__ */ _defineComponent({\n'
        + '  __name: "App",\n'
        + '  setup(__props, { expose }) {\n'
        + '    expose();\n'
        + '    _useCssVars((_ctx) => ({\n'
        + '      "229090c3-sassColor": sassColor.value\n'
        + '    }));\n'
        + '    const color = ref("red");\n'
        + '    const sassColor = ref("#94c9ff");\n'
        + '    return { color, sassColor }\n'
        + '  }\n'
        + '});'
      const parseRes = parserCompiledSfc(mockContent)
      const mgcStr = new MagicString(mockContent)
      const res = injectCSSVarsOnServer(
        [{ value: 'color', has: true, isRef: true }] as any,
        true,
        parseRes,
        mgcStr)
      expect(res.mgcStr.toString()).toContain('color.value')
    })

    test('isScriptSetup=false hasUseCssVars=true', () => {
      const mockContent = 'import { useCssVars as _useCssVars } from \'vue\'\n'
        + 'const __injectCSSVars__ = () => {\n'
        + '_useCssVars(_ctx => ({\n'
        + '  "229090c3-sassColor": (_ctx.sassColor)\n'
        + '}))}\n'
        + 'const __setup__ = __default__.setup\n'
        + '__default__.setup = __setup__\n'
        + '  ? (props, ctx) => { __injectCSSVars__();return __setup__(props, ctx) }\n'
        + '  : __injectCSSVars__\n'
        + '\n'
        + 'const _sfc_main = __default__'

      const parseRes = parserCompiledSfc(mockContent)
      const mgcStr = new MagicString(mockContent)
      const res = injectCSSVarsOnServer(
        [{ value: 'color', has: true, isRef: true }] as any,
        false,
        parseRes,
        mgcStr)
      expect(res.mgcStr.toString()).toContain('_ctx.color')
    })

    test('isScriptSetup=true hasUseCssVars=false', () => {
      const mockContent = 'const _sfc_main = /* @__PURE__ */ _defineComponent({\n'
        + '  __name: "App",\n'
        + '  setup(__props, { expose }) {\n'
        + '    expose();\n'
        + '    const color = ref("red");\n'
        + '    const sassColor = ref("#94c9ff");\n'
        + '    return { color, sassColor }\n'
        + '  }\n'
        + '});'
      const parseRes = parserCompiledSfc(mockContent)
      const mgcStr = new MagicString(mockContent)
      const res = injectCSSVarsOnServer(
        [{ value: 'color', has: false, isRef: true }] as any,
        true,
        parseRes,
        mgcStr)
      expect(res.mgcStr.toString()).toContain('color.value')
      expect(res.mgcStr.toString()).toContain('_useCssVars')
    })
  })

  describe('injectCSSVars', () => {
    test('basic', () => {
      const mockContent = 'const _sfc_main = /* @__PURE__ */ _defineComponent({\n'
        + '  __name: "App",\n'
        + '  setup(__props, { expose }) {\n'
        + '    expose();\n'
        + '    const color = ref("red");\n'
        + '    const sassColor = ref("#94c9ff");\n'
        + '    return { color, sassColor }\n'
        + '  }\n'
        + '});'
      const parseRes = parserCompiledSfc(mockContent)
      const mgcStr = new MagicString(mockContent)
      const res = injectCSSVars(
        [{ value: 'color', has: false, isRef: true }] as any,
        true,
        parseRes,
        mgcStr)
      expect(res.mgcStr.toString()).toContain('color.value')
      expect(res.mgcStr.toString()).toContain('_useCssVars')
    })

    test('vbindVariableList is empty', () => {
      const mockContent = 'const _sfc_main = /* @__PURE__ */ _defineComponent({\n'
        + '  __name: "App",\n'
        + '  setup(__props, { expose }) {\n'
        + '    expose();\n'
        + '    const color = ref("red");\n'
        + '    const sassColor = ref("#94c9ff");\n'
        + '    return { color, sassColor }\n'
        + '  }\n'
        + '});'
      const parseRes = parserCompiledSfc(mockContent)
      const mgcStr = new MagicString(mockContent)
      const res = injectCSSVars(
        [] as any,
        true,
        parseRes,
        mgcStr)
      expect(res.mgcStr.toString()).toContain(mockContent)

      const res2 = injectCSSVars(
        null!,
        true,
        parseRes,
        mgcStr)
      expect(res2.mgcStr.toString()).toContain(mockContent)
    })
  })
})
