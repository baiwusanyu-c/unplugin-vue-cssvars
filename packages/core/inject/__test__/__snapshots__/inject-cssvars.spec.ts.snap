// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inject-cssvars > createCSSVarsObjCode > hasUseCssVars is false 1`] = `
"import { useCssVars as _useCssVars } from \\"vue\\"
const __default__ = {};const __injectCSSVars__ = () => {}

        const __setup__ = __default__.setup
        __default__.setup = __setup__
          ? (props, ctx) => { __injectCSSVars__(); return __setup__(props, ctx) }
            : __injectCSSVars__
            const _sfc_main = __default__
            function _sfc_render(){}"
`;

exports[`inject-cssvars > createCSSVarsObjCode > hasUseCssVars is true 1`] = `
"import { useCssVars as _useCssVars } from 'vue'
const __injectCSSVars__ = () => {
_useCssVars(_ctx => ({
\\"color\\": (_ctx.color),

  \\"229090c3-sassColor\\": (_ctx.sassColor)
}))}
const __setup__ = __default__.setup
__default__.setup = __setup__
  ? (props, ctx) => { __injectCSSVars__();return __setup__(props, ctx) }
  : __injectCSSVars__

const _sfc_main = __default__"
`;

exports[`inject-cssvars > injectUseCssVarsSetup > hasUseCssVars is false 1`] = `
"import { useCssVars as _useCssVars } from \\"vue\\"
function setup() {
     _useCssVars((_ctx) => ({
      \\"1439c43b\\": color.value,
  }));}"
`;

exports[`inject-cssvars > injectUseCssVarsSetup > hasUseCssVars is true 1`] = `
"function setup() {  _useCssVars((_ctx) => ({\\"1439c43b\\": color.value,
    \\"229090c3-sassColor\\": sassColor.value
})); }"
`;
